### DISCUSSION TITLE
table多样化编辑中的问题

### DISCUSSION THREAD

**Discussion by hjingsuper** [layui/layui#1941] [讨论]
根据 #1891 测试后发现的问题

示例：[https://codepen.io/hjingsuper/pen/vYwGMXv](https://codepen.io/hjingsuper/pen/vYwGMXv)

> 1.`table.updateRow(id, opts);`和`obj.update(fields, related);`的参数`related:true`时，好像并没有更新缓存。
```
  // 行单击事件
  table.on("rowDouble(MianTable)", function (obj) {
    var data = obj.data; // 得到当前行数据
    obj.update({ IsEdit: 1 }, true); // 修改行数据
  });
```
操作步骤：
新增行-双击行-选择性别-新增行（此时会重置选择项）
![2024-05-23-16-42-40](https://github.com/layui/layui/assets/36070786/022b3ed7-d556-489f-9f09-b7863e42f71b)

> 2. `obj.update`时不设置`related`参数，使用`renderData`函数更新缓存，此时没有问题。但是这个时候要设置`scrollPos: "fixed"`属性，不然滚动条会回到顶部。或许`renderData`也可以加个参数呢，保持滚动条位置不变。个人想法，只是要注意设置
```
  // 行单击事件
  table.on("rowDouble(MianTable)", function (obj) {
    var data = obj.data; // 得到当前行数据
    obj.update({ IsEdit: 1 });
    //该方法用于重新渲染数据，一般在修改 table.cache 后使用。
    table.renderData("MianTable");
  });
```
![2024-05-23-17-36-36](https://github.com/layui/layui/assets/36070786/67ff0211-a03d-4e68-a990-56392e721406)

**Comment by Sight-wcg** [comment]
> 1.table.updateRow(id, opts);和obj.update(fields, related);的参数related:true时，好像并没有更新缓存。   

详细原因见我之前的回复 https://github.com/layui/layui/discussions/1774#discussioncomment-9077177

```
       var tableViewElem = this.elem.next();
          // 解除 tbSelect 命名空间下的所有 change 事件处理程序
          tableViewElem.off("change.tbSelect");
          // 将 '.select-sex' 元素的 change 事件委托给 tableViewElem, 事件命名空间为 tbSelect
          tableViewElem.on("change.tbSelect", ".select-sex", function () {
            var value = this.value; // 获取选中项 value
            var data = table.getRowData(options.id, this);
            data.sex = value;
          });
```

> 2.或许renderData也可以加个参数呢，保持滚动条位置不变

table.renderData 设计上只用来更新本地的缓存数据，不能更新其他选项，这一块目前不太好动，建议表格初始化时就设置好 `scrollPos: "fixed"`

**Reply by hjingsuper** [reply]
好的，没有问题了，没细看table的属性，缺少设置。感谢
官网文档[多样化编辑](https://layui.dev/docs/2/table/#demo-editable)示例代码应该要改下
